File: sort/quick.go
package sort

//快排
func QuickSort(s []int, low int, high int) {
	//low, high := 0, len(s)
	//pivot := rand.Intn(high)
	if low < high {
		pivot := partition(s, low, high)
		QuickSort(s, low, pivot-1)
		QuickSort(s, pivot+1, high)
	}

}

func partition(s []int, low int, high int) int {
	pivot := s[low]
	for low < high {
		for ; low < high && s[high] >= pivot; high-- {
		}
		s[low] = s[high]
		for ; low < high && s[low] <= pivot; low++ {
		}
		s[high] = s[low]

	}
	s[low] = pivot
	return low
}

File: py/structure/sort/quick.py
# -*- coding:utf-8 -*-
"""
author:
date:
version:
content:
"""
from typing import List


# 快速排序
def quick_sort(ls: List, low: int, high: int) -> None:
    """
    """
    if low < high:
        pivot = partition(ls, low, high)
        quick_sort(ls, low, pivot - 1)
        quick_sort(ls, pivot + 1, high)


def partition(ls: List, low: int, high: int) -> int:
    pivot = ls[low]
    while low < high:
        while low < high and pivot <= ls[high]:
            high -= 1
        ls[low] = ls[high]

        while low < high and pivot >= ls[low]:
            low += 1
        ls[high] = ls[low]

    ls[low] = pivot

    return low


if __name__ == '__main__':
    s = [1, 3, 6, 7, 8, 9, 2, 5, 1]
    quick_sort(s, 0, len(s) - 1)
    print(s)
